Allow audit to log authorization failures

soltysh opened the pull request:
**What this PR does / why we need it**:
This PR extends our current audit mechanism allowing to audit failed login attempts. 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Advanced audit allows logging failed login attempts
```

soltysh committed:
Allow audit to log authorization failures

soltysh commented:
The only remaining part is tests for the new bits.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,196 @@
+		}
+
+		// otherwise, we need to create the event by ourselves and log the auth error
+		// the majority of this code is copied from upstream WithAudit filter
```
s/upstream//

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,196 @@
+		ev.ResponseStatus = &metav1.Status{}
+		ev.ResponseStatus.Code = int32(code)
+		ev.Stage = auditinternal.StageResponseStarted
+		processEvent(a.sink, ev)
```
as we are in another file now, maybe rename this to processAuditEvent

soltysh commented:
@sttts I've added tests and run bazel, ptal

crassirostris commented:
@destijl Could you please take a look?

/cc @tallclair

sttts commented:
lgtm. But two more eyes on it wouldn't harm.

destijl commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,206 @@
+		if err != nil {
+			return basic
+		}
+		cred := strings.SplitN(string(str), ":", 2)
```
This seems to give an unauthenticated attacker complete control over the username field? Potential XSS vector if downstream consumers are displaying the contents?

soltysh replied:
This piece of code is taken directly from basic auth, so the same problem applies there. If I'm able to address other comment of yours, this should not be a problem anymore.

sttts replied:
/cc @liggitt

sttts replied:
Any user can fake a user-name and this name is extracted here to be added to the audit event on authn failure. This by itself is expected and intended. But I agree that allowing any kind of binary data might open up an attack vector on buggy consumers. What is best practice here?

liggitt replied:
same vector applies to resource name parsed from the url, right?

sttts replied:
For resource names at least the charset is restricted.

liggitt replied:
> For resource names at least the charset is restricted.
not by request parsing, the restrictions are per resource

destijl commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// getUsername returns username or information on the authn method being used.
+func getUsername(req *http.Request) string {
```
AIUI we get an authentication failure, then handover the failed request to this code to try and extract some context from the request. Is there something stopping us logging a more meaningful failure at the actual authentication failure time? Seems like the point where we determine the password/cert/etc is bad is the place where we should create the log.

This hard-codes some assumptions about the auth methods we accept which would seem to better belong in a single library (or libraries, one for each method) so that we can be careful about how we handle this untrusted input and reason about code changes needed to deprecate authentication methods if necessary.

soltysh replied:
Hrm... I'm not a big fan of it either, but it was the first thing I could think of. I'll dig into the code once more and see how we can do a better job here.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,206 @@
+
+	// check basic auth
+	const basicScheme string = "Basic "
+	if strings.HasPrefix(auth, basicScheme) {
```
https://golang.org/pkg/net/http/#Request.BasicAuth ?

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	// other tokens
+	token := strings.TrimSpace(req.URL.Query().Get("access_token"))
```
remove this, this method is not supported

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,206 @@
+		if len(cred) < 2 {
+			return basic
+		}
+		return cred[0]
```
I agree that putting unverified data about users in audit logs is fraught

soltysh commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,187 @@
+		}
+
+		// if there already exists an audit event in the context we don't need to do anything
+		if ae := request.AuditEventFrom(ctx); ae != nil {
```
I left this check here, but it shouldn't be needed, since we hook in failed authn specifically.

soltysh commented:
@liggitt @sttts @destijl I've addressed your concerns, can I get a quick review on this one? 
The only remaining thing I'd like to do here, is to inject auth method name in `user.Extras`, b/c in majority of cases we're not getting username, but we could at least point to which methods were tried, wdyt?

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:
```diff
@@ -936,21 +936,24 @@ func TestAuditJson(t *testing.T) {
 func (m *fakeRequestContextMapper) Get(req *http.Request) (request.Context, bool) {
-	ctx := request.NewContext()
+	if m.ctx == nil {
+		m.ctx = request.NewContext()
```
this makes fakeRequestContextMapper non-thread safe

soltysh replied:
But this is just test, for that particular usecase. I doubt `fakeRequestContextMapper` will be use anywhere else. Same for the next call, as well.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go:
```diff
@@ -936,21 +936,24 @@ func TestAuditJson(t *testing.T) {
 	info, err := resolver.NewRequestInfo(req)
 	if err == nil {
-		ctx = request.WithRequestInfo(ctx, info)
+		m.ctx = request.WithRequestInfo(m.ctx, info)
```
and this makes it not safe to be called multiple times

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:
```diff
@@ -63,6 +63,9 @@ func WithAuthentication(handler http.Handler, mapper genericapirequest.RequestCo
 					glog.Errorf("Unable to authenticate the request due to an error: %v", err)
 				}
+				if auditFailedAuthn != nil {
+					w = auditFailedAuthn.Decorate(w, user, req)
```
why are you passing user? when `err != nil || !ok`, `user` is unusable

liggitt replied:
and if you don't need user, then `auditFailedAuthn.Decorate` has the same args as `http.Handler`, and you can simple pass a wrapped failed http.Handler in here that logs, then handles

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -48,37 +48,17 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 
-		attribs, err := GetAuthorizerAttributes(ctx)
+		ev, err := createAuditEvent(ctx, policy, requestContextMapper, req, w)
 		if err != nil {
```
shouldn't this still lead to http 500?

sttts replied:
Ic, createAuditEvent writes the error. I would prefer to see it here.

sttts replied:
or at least a good comment here.

soltysh replied:
I'll move the error handling up, it'll be more explicit.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -116,20 +96,52 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
+		return nil, err
+	}
+
+	ctx = request.WithAuditEvent(ctx, ev)
```
this is unexpected in a `create*` function as a side-effect.

sttts replied:
either move that our or call the func createAuditEventAndAttachToCtx

soltysh replied:
I moved context extraction also to that method and went with `createAuditEventAndAttachToContext`.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,92 @@
+			return
+		}
+
+		ev, err := createAuditEvent(ctx, policy, requestContextMapper, req, w)
```
so this attaches an event to the context. The `WithAudit` filter will override that. Correct?

soltysh replied:
See above.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,92 @@
+		ev.ResponseStatus.Message = getAuthMethods(req)
+		ev.Stage = auditinternal.StageResponseStarted
+
+		rw := decorateResponseWriter(w, ev, sink)
```
This makes the rw.WriteHeader to process an event. The `WithAudit` also decorates. So if an endpoint handler writes the header, the `WithAudit` decorator will send an event to the audit backend, then your new outer decorator will the same. Moreover, you `ev` here is another one than the one of `WithAudit`. So yours will be pretty empty.

soltysh replied:
No double decoration will happen. he reason for that is that `WithFailedAuthenticationAudit` decorates *only* the failed handler that is passed to `WithAuthentication`. At that point no request will ever reach regular `WithAudit` because the processing is passed through a different handler chain.

sttts replied:
got it. Please call the parameter `failedHandler` here. Then it's clearer.

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,89 @@
+
+// WithFailedAuthenticationAudit decorates a http.Handler with a fallback audit,
+// logging information only when the original one did was not triggered.
+// This needs to be used with WithAuditTriggeredMarker, which wraps the original
```
I can't find "WithAuditTriggeredMarker" in other places.
Maybe this doc string is out of date?

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit_test.go:
```diff
@@ -0,0 +1,124 @@
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit/policy"
+	// import to call webhook's init() function to register audit.Event to schema
+	_ "k8s.io/apiserver/plugin/pkg/audit/webhook"
```
I have tested that it's not necessary to import this in this unit test.

soltysh commented:
Updated with addressed comments from @CaoShuFeng

sttts commented:
lgtm after the `failedHandler` and @CaoShuFeng's comment are fixed.

sttts commented:
/approve no-issue
/lgtm

soltysh commented:
/retest

CaoShuFeng commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go:
```diff
@@ -0,0 +1,87 @@
+
+		ev.ResponseStatus = &metav1.Status{}
+		ev.ResponseStatus.Message = getAuthMethods(req)
+		ev.Stage = auditinternal.StageResponseStarted
```
I am OK with `auditinternal.StageResponseStarted` here.
But why `StageResponseStarted` is better than `StageRequestReceived`?
